Temperature pumps can now transfer heat against the gradient. Changes thermomachine power consumption to be perfectly linear when it doesn't reach the exponential threshold. Adds define for heat power conversion.

Pickle-Coding opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Temperature pumps can now transfer heat against the gradient. When on, it will equalize the temperature, then transfer heat as if the input node is transferring heat with something colder than it is. Thermomachine power consumption is now linear if it doesn't reach the exponential power consumption threshold. Removes thermomachine power conversion define, and instead make a heat power conversion define.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
The operation of the temperature pump was too similar to the heat exchangers. This change will allow more different ways for people to interact with heat. Thermomachine power consumption having diminishing returns when it didn't reach exponential power consumption threshold was weird.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
balance: Temperature pumps can now pump heat.
balance: Thermomachine power consumption has been changed so it is a linear function when it doesn't reach the exponential power consumption threshold. It will consume 0.0001 joules of power per joule of heat changed, if the heat exchanged doesn't surpass 5GJ.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Pickle-Coding committed:
Temperature pump change.

Changes functionality of temperature pump. It will now exchange heat and equalise temperature between the two nodes, then, if powered, pump heat. It will consume power based on the heat transfer.

RandomGamer123 commented:
Temperature pumps being togglable was one nice thing that differed them from heat exchangers, by making temperature pumps always equalize heat, they actually become more similar to heat exchangers.

Pickle-Coding commented:
Maybe I could make them only do the equalise while powered.

Pickle-Coding commented:
Actually this is way too OP for something that can get printed by the RPD for free. It allows you to cool nitrogen or whatever below 30 Kelvin roundstart if you use an equal amount of plasma as the thermal reservoir. I'll reopen this or make a new PR if I can balance it better.

Pickle-Coding committed:
Okay.

The on check will occur before heat exchange, so it will no longer exchange heat if it's off.

Adds define for heat and power conversion, and makes both the thermomachine and temperature pump use it, instead of their own defines.

Changes thermomachine power consumption to be linear, and removes power consumption limit.

Changes temperature pump conduction behavior to treat the temperature delta to be the delta of 15% of its temperature, if not the delta of TCMB.

Pickle-Coding commented:
I have changed the functionality to make it much harder to reach extremely low temperatures from room temperature gas, so hopefully they won't be better than thermomachines. I also made it to not conduct heat if it's not powered.

Pickle-Coding committed:
Change.

Changes thermomachine power usage to have the exponential power usage after a certain threshold. Changes it slightly so the function is linear if it hasn't met the threshold yet, and removes power consumption limit.

LemonInTheDark commented on code/modules/atmospherics/machinery/components/binary_devices/temperature_pump.dm:
```diff
@@ -43,17 +43,26 @@
-		remove_input.temperature = max(remove_input.temperature - (cooling_heat_amount / input_capacity), TCMB)
-		update_parents()
+	//Pump heat against the gradient.
+	var/heat_amount = CALCULATE_CONDUCTION_ENERGY((thermal_equilibrium - max(thermal_equilibrium * 0.15, TCMB)) * heat_transfer_rate * 0.01, input_capacity, output_capacity)
```
why the double () wrap? should be done on the define's end

Pickle-Coding replied:
So that the energy that gets transferred gets multiplied by the heat transfer rate, rather than the fake temperature delta or whatever.

LemonInTheDark commented:
I haven't checked in detail but my gut is the pumping here is gonna cause troubles, being able to suck energy feels like it'll be pain, and make balancing temperatures suffering

LemonInTheDark commented:
?

Pickle-Coding commented:
I forgot about the gas turbine heat conversion, and I wanna do this PR at a different time that isn't rushed.
